libslide/src/scanner.rs

```diff
@@ -0,0 +1,209 @@
+mod types;
+use types::TokenType;
+use types::Token;
+
+pub struct Scanner {
+    input: String, 
+    pub output: Vec<Token>
+}
+
+impl Scanner {
+    // instantiate a new scanner
+    pub fn new<T: Into<String>>(input: T) -> Scanner {
+        Scanner{
+            input: input.into(),
+            output: Vec::new()
+        }
+    }
+
+    // matches token with symbol and creates it: private helper function 
+   fn create_symbol_token(c: char) -> Token{
+        let t = match c {
+            '+' => TokenType::Plus,
+            '-' => TokenType::Minus,
+            '*' => TokenType::Mult,
+            '/' => TokenType::Div,
+            '%' => TokenType::Mod,
+            '^' => TokenType::Exp,
+            '=' => TokenType::Equal,
+            '(' => TokenType::OpenParen, 
+            ')' => TokenType::CloseParen, 
+            '[' => TokenType::OpenBracket,
+            ']' => TokenType::CloseBracket,
+             _  => TokenType::Invalid(c.to_string())
+        };
+        let ret = Token{token: t};
+        return ret;
+   }
+
+    // iterates through any digits to create a token of that value
+    fn iterate_digit(&mut self,mut i: usize) -> (Token, usize){
+        let mut int_str = "".to_owned();
+        let mut dec_str = ".".to_owned();
+        let ret: Token;
+        // iterate through integer part
+        while i < self.input.chars().count() && (self.input.as_bytes()[i] as char).is_digit(10){
+            int_str.push(self.input.as_bytes()[i] as char);
+            i += 1;
+        }
+        // iterate through decimal
+        if i < self.input.chars().count() && (self.input.as_bytes()[i] as char)== '.'{
+            i += 1;
+            while i < self.input.chars().count() && (self.input.as_bytes()[i] as char).is_digit(10){
+                dec_str.push(self.input.as_bytes()[i] as char);
+                i += 1;
+            }
+            int_str.push_str(&dec_str);
+            // turn integer and decmial strings into token
+            ret = Token{token: TokenType::Float(int_str.parse::<f64>().unwrap())}
+        }
+        else{
+            // turn integer string into token and default the float
+            ret = Token{token:TokenType::Int(int_str.parse::<i64>().unwrap())}
+        }
+        return (ret, i);
+    }
+
+    pub fn scan(&mut self){
+        let mut i: usize = 0;
+        let mut c: char;
+        let mut t: Token;
+        let mut tuple: (Token, usize);
+        // iterate through string
+        while i < self.input.chars().count() {
+            // ignore whitespace
+            if !((self.input.as_bytes()[i] as char).is_whitespace()) {
```

Comment:
Can you not do
```suggestion
            if !((self.input[i]).is_whitespace()) {
```
?

Reply:
no

Reply:
strings cannot be indexed as you'd expect in rust. They have to coverted to bytes. It is O(1) though so I think its ok

Reply:
Oh that's weird. In this case, shouldn't we convert to as_bytes before we start the lexing?

Reply:
thatd probably be a cleaner way to do it

Reply:
so changing everything to as_bytes gets confusing
Basically as bytes returns a utf8 array. However, setting input to a utf8 array results in an unexpected size error. Therefore, we should write input as a reference to a utf8 array. However, this then gets into lifetime issues in structs. From there we would need to declare an explicit lifetime of our struct scan which seems very complex to just simplify as_bytes when as_bytes is O(1)